### PR TITLE
expression, ranger: three optimizations to speedup some specific query's opt time

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Three optimizations in this pr.
- Pushdown Not don't call NewFunction if not new function is created
- FoldConstant doesn't need to traversal the whole expr tree.
- When merge Ranges, we don't need to new slice again.

This three can reduce the SQL's optimizing(or compling) time significantly(can be ***13%*** compared with the master branch) if it contains a large OR tree. (i.e. where a = ... or a = ... or a = ... or ... or ...)

### What is changed and how it works?

The first one and the third one is easy to understand.
Some explanation about the second one:
- We have two function call currently: `NewFunction` and `NewFunctionBase`
- `NewFunctionBase` banned the foldConstant. Which means that the expr should not be folded anymore in any case.
- `NewFunction` fold the constant every time the function is built. So if the expr is generated by `NewFunction`. It must be totally folded.
- Mege the two case. You can see that the expr's children don't needs to check whether can fold or not since its child must arrive its final state unless it's craeted from the other function.

### Check List 

Tests 

 - Exsiting Unit test

Code changes

 - Has exported function/method change

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: expression/util.go [hunk 1]
```diff
@@ -288,46 +290,55 @@ var symmetricOp = map[opcode.Op]opcode.Op{
...
+		flag = changed || flag
+		newExprs = append(newExprs, newExpr)
 	}
-	return newExprs
+	return newExprs, flag
 }
 
 // PushDownNot pushes the `not` function down to the expression's arguments.
-func PushDownNot(ctx sessionctx.Context, expr Expression, not bool) Expression {
+func PushDownNot(ctx sessionctx.Context, expr Expression, not bool) (Expression, bool) {
```
Comment (lamxTyler) [inline expression/util.go]: Can we rewrite it like `FoldConstant` that only returns expression, but `foldConstant` would return other values, it could save many unnecessary changes.

Comment (winoros) [reply expression/util.go]: Not very easy for now.
`FoldConstant` is called everytime when the expr is built so we don't need to traverse the whole tree.
But `PushDownNot` is not, so we know the expr_1 needs to be generated as a new expr only when the subtree is checked. Thus, we need a extra returned value to tell us the status of the subtree.
If we want to ingore it. We need to move it inside the `NewFunction`.

Comment (lamxTyler) [reply expression/util.go]: Oh, I mean we can rewrite it like
```
func PushDownNot(...) Expression
{
   expr, _ := pushDownNot(...)
   return expr
}
```
so people that uses `PushDownNot` do not need to care about the useless value.

Comment (eurekaka) [reply expression/util.go]: +1

Comment (winoros) [reply expression/util.go]: fixed
